af/fastpath/fastpath/core.py

```diff
@@ -0,0 +1,574 @@
+
+    measurement["test_keys"]["requests"] = None  # TODO: delete
+
+
+@metrics.timer("load_s3_reports")
+def load_s3_reports(day) -> dict:
+    # TODO: move this into s3feeder
+    t0 = time.time()
+    path = conf.s3cachedir / str(day)
+    log.info("Scanning %s", path.absolute())
+    files = []
+    with os.scandir(path) as d:
+        for e in d:
+            if not e.is_file():
+                continue
+            if e.name == "index.json.gz":
+                continue
+            files.append(e)
+
+    fcnt = 0
+    for e in sorted(files, key=lambda f: f.name):
+        log.debug("Ingesting %s", e.name)
+        fn = os.path.join(path, e.name)
+        fcnt += 1
+        for measurement_tup in s3feeder.load_multiple(fn):
+            yield measurement_tup
+
+        remaining = (time.time() - t0) * (len(files) - fcnt) / fcnt
+        metrics.gauge("load_s3_reports_eta", remaining)
+        metrics.gauge("load_s3_reports_remaining_files", len(files) - fcnt)
+        remaining = timedelta(seconds=remaining)
+        log.info("load_s3_reports remaining time: %s", remaining)
+
+
+def prepare_for_json_normalize(report):
+    try:
+        d = report["test_keys"]["control"]["tcp_connect"]
+        d = {n: i for n, i in enumerate(d.items())}
+        report["test_keys"]["control"]["tcp_connect"] = tuple(d.items())
+    except KeyError:
+        pass
+
+    try:
+        h = report["test_keys"]["control"]["http_request"]["headers"]
+        report["test_keys"]["control"]["http_request"]["headers"] = tuple(h.items())
+    except KeyError:
+        pass
+
+
+def fetch_measurements(start_day, end_day) -> dict:
+    """Fetch measurements from S3 and the collectors
+    """
+    # no --start-day or --end-day   -> Run over SSH
+    # --start-day                   -> Run on old cans, then over SSH
+    # --start-day and --end-day     -> Run on old cans
+    # --start-day and --end-day  with the same date     -> NOP
+    today = date.today()
+    if start_day and start_day < today:
+        # TODO: fetch day N+1 while processing day N
+        log.info("Fetching older cans from S3")
+        day = start_day
+        while day < (end_day or today):
+            log.info("Processing %s", day)
+            s3feeder.fetch_cans_for_a_day_with_cache(conf, day)
+            for measurement_tup in load_s3_reports(day):
+                yield measurement_tup
+
+            day += timedelta(days=1)
+
+        if end_day:
+            return
+
+    ## Fetch measurements from collectors: backlog and then realtime ##
+    log.info("Starting fetching over SSH")
+    for measurement_tup in sshfeeder.feed_measurements_from_collectors(conf):
+        yield measurement_tup
+
+
+@metrics.timer("match_fingerprints")
+def match_fingerprints(measurement):
+    """Match fingerprints against HTTP headers and bodies.
+    """
+    # TODO: apply only on web_connectivity
+    msm_cc = measurement["probe_cc"]
+    if msm_cc not in fingerprints:
```

Comment:
My understanding is that you use `ZZ` in the fingerprints dict to indicate ANY country. In the case in which there aren't country specific fingerprints maybe we still want to run the fingerprint check against the fingerprints of `ZZ` and I think this check will skip that.

Reply:
There was a TODO in setup_fingerprints for this. Updated to handle ZZ.